### PR TITLE
Rewrite VirtualEnvGenerator tests

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: omit
Docs: omit

All the valuable code is not mine, I have just copied it from #4291. Kudos to @ytimenkov 

What's in this PR?
 * No changes to production codebase, only testing
 * Unit testing for _env_ generators
 * **Actual testing of VirtualEnvGenerator** in `conans/test/functional/generators/virtualenv_test.py `
 * **Actual testing of RunEnvironment** in `conans/test/integration/run_environment_test.py`
 * Some improvements to the `MockConanfile` needed for these tests to run

I've introduced some `FIXME` with things that don't work as expected, some of them are just typos. It is not the objective of this PR to question if it is working well or not, the idea is to show how it works right now.

📣  If you think I should add any extra testing or any behavior or casuistic that it is not covered right now, please comment below and I'll do. If you think that there are any repeated test that can be removed, say it. Thanks!

This is the first step for https://github.com/conan-io/conan/issues/5821

### CODE REVIEW

**Comment by memsharded** [comment]
I am thinking that if a new paradigm of virtualenvs is to be deployed, maybe the testing system could already be defined around it. Like the generation of "deactivate" scripts at generation time is broken by definition: The deactivate information can only be generated at "activate" time.

**Comment by ytimenkov** [comment]
> the generation of "deactivate" scripts at generation time is broken by definition: The deactivate information can only be generated at "activate" time.

That is exactly how/why I started this. :D

Btw, yesterday my colleague got a problem because of this: deactivate script captured LD_Library_path and he got really weird errors from dynamic loader.

I also put my thoughts into #4604

**Comment by jgsogo** [comment]
First thing first: test what we are currently doing, these new tests (checking the actual environment) are a huge step forward compared to what we are doing now. Next step is to modify behavior and see the differences with the current one.

For sure the one you are pointing out is very important, and it will show up as soon as we introduce a test with nested _virtualenvs_. I'll take care of it, this journey has just started 🚀

**Comment by jgsogo** [comment]
Yes, some of the FIXME are really easy ones indeed. When merged, I'll open issues (maybe grouping some of them, and probably as `good first issues` + `hacktoberfest` as they should be really small changes targeting already provided tests).

File: conans/test/functional/generators/virtualenv_test.py [hunk 1]
```diff
@@ -1,196 +1,275 @@
...
+
+        # Consistency checks
+        self.assertFalse(stderr, "Running shell resulted in error, output:\n%s" % stdout)
+
+        env_before = _load_env_file(os.path.join(self.test_folder, self.env_before))
+        env_after = _load_env_file(os.path.join(self.test_folder, self.env_after))
+        if platform.system() == "Darwin":
+            env_after.pop(six.u("PS1"), None)  # TODO: FIXME: Needed for the test to pass
+            env_after.pop("PS1", None)  # TODO: FIXME: Needed for the test to pass
+        self.assertDictEqual(env_before, env_after)  # Environment restored incorrectly
```
Comment (memsharded) [inline conans/test/functional/generators/virtualenv_test.py]: incorrectly => correctly

**Comment by memsharded** [comment]
this is great! Thanks @jgsogo and @ytimenkov for this contribution.

**Pull request merged**